netket/models/autoreg.py

```diff
@@ -102,12 +103,73 @@ def __call__(self, inputs: Array) -> Array:
         if inputs.ndim == 1:
             inputs = jnp.expand_dims(inputs, axis=0)
 
-        idx = (inputs + self.hilbert_local_size - 1) / 2
+        idx = (inputs + self.hilbert.local_size - 1) / 2
         idx = jnp.asarray(idx, jnp.int64)
-        idx = jnp.expand_dims(idx, axis=2)
+        idx = jnp.expand_dims(idx, axis=-1)
 
         p, _ = self.conditionals(inputs, None)
         p = jnp.take_along_axis(p, idx, axis=1)
 
-        log_psi = 1 / 2 * jnp.log(p + self.eps).sum(axis=(1, 2))
+        log_psi = 1 / 2 * jnp.log(p + self.eps)
+        log_psi = log_psi.reshape((inputs.shape[0], -1)).sum(axis=1)
         return log_psi
+
+
+class ARNNConv1D(ARNN):
+    """Autoregressive neural network with 1D convolution layers."""
+
+    hilbert: CustomHilbert
+    """the discrete Hilbert space."""
+    layers: int
+    """number of layers."""
+    features: Union[Iterable[int], int]
+    """number of features in each layer. If a single number is given,
+    all layers except the last one will have the same number of features."""
+    kernel_size: int
+    """length of the convolutional kernel."""
+    kernel_dilation: int = 1
+    """dilation factor of the convolution kernel (default: 1)."""
+    activation: Callable[[Array], Array] = nn.silu
+    """the nonlinear activation function between hidden layers (default: silu)."""
+    use_bias: bool = True
+    """whether to add a bias to the output (default: True)."""
+    dtype: DType = jnp.float64
+    """the dtype of the weights (default: float64)."""
+    precision: Any = None
+    """numerical precision of the computation, see `jax.lax.Precision`for details."""
+    kernel_init: NNInitFunc = lecun_normal()
+    """initializer for the weights."""
+    bias_init: NNInitFunc = zeros
+    """initializer for the biases."""
+    eps: float = 1e-7
+    """a small number to avoid numerical instability."""
+
+    def setup(self):
+        if isinstance(self.features, int):
+            features = [self.features] * (self.layers - 1) + [self.hilbert.local_size]
+        else:
+            features = self.features
+        assert len(features) == self.layers
+        assert features[-1] == self.hilbert.local_size
+
+        self._layers = [
+            MaskedConv1D(
+                features=features[i],
+                kernel_size=self.kernel_size,
+                kernel_dilation=self.kernel_dilation,
+                exclusive=(i == 0),
+                use_bias=self.use_bias,
+                dtype=self.dtype,
+                precision=self.precision,
+                kernel_init=self.kernel_init,
+                bias_init=self.bias_init,
+            )
+            for i in range(self.layers)
+        ]
+
+    def conditionals(self, inputs: Array, cache: PyTree) -> Tuple[Array, PyTree]:
+        return ARNNDense.conditionals(self, inputs, cache)
```

Comment:
is this ok? I don't understand what happens in here...
You are not even creating an ARNNDense field in setup..

Reply:
I am 'borrowing' the methods `conditionals` and `__call__` from `ARNNDense` to avoid some duplicated code. `ARNNDense.conditionals` is just a function, and its first argument can be any object. I admit that it's not a common practice, and may conflict with Flax's black magics of hiding members of `Module` if there are further changes...

Another way may be to implement a base class of ARNN with sequential layers, and pass `_layers` as a property when init. That will be different from the current interface of `GCNN`.

Reply:
ok. I'll think a moment about this

Reply:
When I read other people's implementations of neural networks with modular structures, I think there're two common practices:

The first is to pass only some hyperparameters to the init of the parent network, and construct the child networks inside the init.

The second is to construct the child networks outside, and pass them to the init of the parent network.

The second way is more composable, but it leaks more labor and implementation details to users.

Reply:
Ah, I see! So the correct way to write this is, IMO, to define a free function 
``` 
def arnn_conditionals(model, inputs, cache):
   x = jnp.expand_dims(inputs, axis=-1)
   ...    
   return ...
```

and then inside of `ARNNConv1D.conditional` and `ARNNDense.conditional` call this function.
Like this is more clear that they share the implementation.